### ISSUE TITLE
Cast train features to float32 in-place after rank normalization to fix OOM crash and halve memory footprint

### ISSUE THREAD

**Issue by mikeycdavis** [mikeycdavis/Numerai#76]
## Problem

The tournament pipeline (\`train-top5.py\`) OOM-crashed twice on 2026-03-20 and once on 2026-03-23 with the following error:

```
numpy._core._exceptions._ArrayMemoryError: Unable to allocate 56.2 GiB
  for an array with shape (2748, 2746268) and data type float64
```

The crash occurs inside \`train_target()\` at:

```python
X = np.array(train[feature_cols].values, dtype=np.float32)   # line 537
```

Pandas' internal \`_interleave\` allocates a **float64** scratch array of the full DataFrame shape before numpy converts to float32. With 2748 features × ~2.75M rows × 8 bytes = 56.2 GiB, this exceeds the 64GB physical limit when other processes are also resident.

## Root Cause

After rank normalization (line 182–185), \`train[feature_cols]\` stores float64 values in [0, 1]. Rank normalization outputs are exact to float32 precision — the float64 precision is entirely wasted, and it doubles the persistent memory footprint of the training DataFrame for the rest of the script.

## Fix

Add one line immediately after rank normalization to cast in-place:

**\`train-top5.py\` — after line 185:**

```python
print("Applying per-era rank normalization to features...")
train[feature_cols] = (
    train.groupby(ERA_COL)[feature_cols]
    .rank(pct=True)
)
# Cast to float32 in-place: rank outputs are [0,1] and need no float64 precision.
# This halves the persistent feature memory (~28GB instead of ~56GB) and eliminates
# the float64 scratch allocation inside train_target() that caused OOM on 2026-03-20.
train[feature_cols] = train[feature_cols].astype(np.float32)
```

With this change, \`train[feature_cols].values\` inside \`train_target()\` returns a float32 array directly — no scratch allocation needed.

## Memory Impact

| | float64 (current) | float32 (after fix) |
|---|---|---|
| \`train\` DataFrame features | ~56 GB | ~28 GB |
| Peak during \`.values\` call | ~112 GB (2×) | ~28 GB (no scratch) |
| Headroom for OS + other models | ~8 GB | ~36 GB |

## Validation Diagnostics

The same fix should be applied to \`validation\` after it is loaded (line 681), for the same reason — validation Sharpe computation and feature exposure calculation read \`validation[feature_cols]\` multiple times:

```python
# After loading validation.parquet:
validation[feature_cols] = validation[feature_cols].astype(np.float32)
```

## Notes

- No change to model quality — rank normalization produces values in \`[0, 1]\` that are identical in float32 and float64
- No CI changes required — the synthetic test data is already small enough that the float64 allocation never exceeds memory
- This fix should be applied before the GPU XGBoost/CatBoost issues (#74, #75) since it reduces baseline memory pressure and makes headroom calculations more predictable
